polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/validation/GradeRequestValidator.java

```
@@ -0,0 +1,140 @@
+    private static final String GRADABLE_EVENT_NOT_IN_COURSE = "GradableEvent is not in this course";
+    private static final String NO_CHEST_IDS_AND_XP = "Either chestIds or xp are required";
+    private static final String TOO_MANY_CHESTS = "Maximum value of chests (%d) with id %d has already been assigned to this grade";
+
+
+    private final CourseGroupRepository courseGroupRepository;
+    private final GradableEventRepository gradableEventRepository;
+
+
+    public void validate(GradeRequestDto request, GradableEvent<?> event, Instructor instructor) {
+        CourseGroup courseGroup = getCourseGroup(request.getCourseGroupId());
+
+        List<ValidationError> errors = Stream.of(
+                        validateInstructorInGroup(instructor, courseGroup),
+                        validateXpOrChestIdsPresent(request),
+                        validateMaxXpNotExceeded(request, event),
+                        validateGradableEventInSameCourseAsCourseGroup(event, courseGroup)
+                )
+
+                .flatMap(Optional::stream)
+                .toList();
+
+
+        if (!errors.isEmpty()) {
+            throw new ValidationException(
+                    ValidationExceptionResponse.builder()
+                            .title(INVALID_REQUEST)
+                            .validationErrors(errors)
+                            .build());
+        }
+    }
+
+    public void validateChestCount(Long chestId, Grade grade) {
+        long assignedChestsCount = Optional.ofNullable(grade.getAssignedChests())
+                .orElse(Collections.emptySet())
+                .stream()
+                .filter(assignedChest -> assignedChest.getChest().getId().equals(chestId))
+                .count();
+
+        Optional<Integer> maxChests = gradableEventRepository.findMaxChestsByGradableEventIdAndChestId(
+                grade.getGradableEvent().getId(),
+                chestId
+        );
```

This field cannot be null (defined by the model). If we get empty optional, the request itself probably was not valid, am I right? Maybe we should throw an error here, instead of `orElse(0)` below? I think that this is even the error we encountered at today's presentation.